lib/mayaUsd/utils/converter.h

```diff
@@ -0,0 +1,331 @@
+#include <maya/MMatrixArray.h>
+#include <maya/MObject.h>
+#include <maya/MPlug.h>
+#include <maya/MDGModifier.h>
+
+#include <mayaUsd/base/api.h>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+MAYAUSD_NS_DEF {
+    
+struct MAYAUSD_CORE_PUBLIC ConverterArgs final
+{
+    UsdTimeCode _timeCode{UsdTimeCode::Default()};
+    bool        _doGammaCorrection{true};
+};
+    
+class MAYAUSD_CORE_PUBLIC Converter final
+{
+public:
+    using MPlugToUsdAttrFn
+        = std::add_pointer<void(const MPlug&, UsdAttribute&, const ConverterArgs&)>::type;
+    using UsdAttrToMPlugFn
+        = std::add_pointer<void(const UsdAttribute&, MPlug&, const ConverterArgs&)>::type;
+    using UsdAttrToMDGModifierFn
+        = std::add_pointer<void(const UsdAttribute&, MPlug&, MDGModifier&, const ConverterArgs&)>::type;
+    using MPlugToVtValueFn
+        = std::add_pointer<void(const MPlug&, VtValue&, const ConverterArgs&)>::type;
+    using VtValueToMPlugFn
+        = std::add_pointer<void(const VtValue&, MPlug&, const ConverterArgs&)>::type;
+    using VtValueToMDGModifierFn
+        = std::add_pointer<void(const VtValue&, MPlug&, MDGModifier&, const ConverterArgs&)>::type;
+    
+    using MDataHandleToUsdAttrFn
+        = std::add_pointer<void(const MDataHandle&, UsdAttribute&, const ConverterArgs&)>::type;
+    using UsdAttrToMDataHandleFn
+        = std::add_pointer<void(const UsdAttribute&, MDataHandle&, const ConverterArgs&)>::type;
+    using MDataHandleToVtValueFn
+        = std::add_pointer<void(const MDataHandle&, VtValue&, const ConverterArgs&)>::type;
+    using VtValueToMDataHandleFn
+        = std::add_pointer<void(const VtValue&, MDataHandle&, const ConverterArgs&)>::type;
+    
+    Converter(const SdfValueTypeName& typeName, MPlugToUsdAttrFn plugToAttr,
+        UsdAttrToMPlugFn attrToPlug, UsdAttrToMDGModifierFn attrToModifier,
+        MPlugToVtValueFn plugToVtValue, VtValueToMPlugFn vtValueToPlug,
+        VtValueToMDGModifierFn vtValueToModifier,
+        MDataHandleToUsdAttrFn handleToAttr,
+        UsdAttrToMDataHandleFn attrToHandle,
+        MDataHandleToVtValueFn handleToVtValue,
+        VtValueToMDataHandleFn vtValueoHandle) : _typeName(typeName),
+        _plugToAttr(plugToAttr), _attrToPlug(attrToPlug),
+        _attrToModifier(attrToModifier), _plugToVtValue { plugToVtValue },
+        _vtValueToPlug { vtValueToPlug }, _vtValueToModifier(vtValueToModifier),
+        _handleToAttr(handleToAttr), _attrToHandle(attrToHandle),
+        _handleToVtValue { handleToVtValue }, _vtValueoHandle { vtValueoHandle }
+    {}
+    
+    Converter() = delete;
+    ~Converter() = default;
+    
+    static const Converter* find(const SdfValueTypeName& typeName, bool isArrayPlug);
+    static const Converter* find(const MPlug& plug, const UsdAttribute& attr);
+    
+    const SdfValueTypeName& type() const
+    { return _typeName; }
+    
+    bool validate(const MPlug& plug, const UsdAttribute& usdAttr, const bool translateMayaDoubleToUsdSinglePrecision) const
```

Comment:
Following the thread from above, maybe this could take a `ConverterArgs` as its last parameter, and use a default-constructed one to make it an optional param?

Reply:
Good point!

Reply:
Looking at it again now. We will need a design for this "double" conversion (i.e. one between maya and usd types, second to single precision). It won't happen as part of this PR, so I'm going to leave this method as is.